lib/cn/csched_sp3_work.c

```diff
@@ -406,33 +450,11 @@ sp3_work_leaf_len(
             }
 
             *rule = CN_CR_LSHORT_LW;
-            runlen = runlen_min;
-        }
-
-
-        /* If the resulting kvset would exceed max_vgroups, convert
-         * the k-compaction to a kv-compaction.
-         */
-        if (runlen > 0) {
-            uint vgroups = 0;
-
-            le = *mark;
-
-            for (uint i = 0; i < runlen; i++) {
-                vgroups += kvset_get_vgroups(le->le_kvset);
-
-                if (vgroups > thresh->max_vgroups) {
-                    *action = CN_ACTION_COMPACT_KV;
-                    *rule = CN_CR_VGMAX;
-                    break;
-                }
-
-                le = list_prev_entry(le, le_link);
-            }
+            return runlen_min;
         }
     }
 
-    return runlen;
+    return 0;
```

Comment:
The above refactoring was necessary to eliminate a bug in which we were performing hordes of 2-kvset kcompactions (because runlen < runlen_min slipped out the bottom of the function when it should have been zero).
Also, I eliminated a big chunk of code that was concerned about spills.